RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitConverter.cc

```diff
@@ -185,9 +189,34 @@ namespace cms {
 
     const edmNew::DetSetVector<SiPixelCluster>& input = *inputhandle;
 
-    edmNew::DetSetVector<SiPixelCluster>::const_iterator DSViter = input.begin();
+    // fill cluster arrays
+    auto hmsp = std::make_unique<uint32_t[]>(gpuClustering::MaxNumModules + 1);
+    auto hitsModuleStart = hmsp.get();
+    std::array<uint32_t, gpuClustering::MaxNumModules + 1> clusInModule{};
+    for (auto DSViter = input.begin(); DSViter != input.end(); DSViter++) {
+      unsigned int detid = DSViter->detId();
+      DetId detIdObject(detid);
+      const GeomDetUnit* genericDet = geom.idToDetUnit(detIdObject);
+      auto gind = genericDet->index();
+      // FIXME to be changed to support Phase2
+      if (gind >= int(gpuClustering::MaxNumModules))
+        continue;
+      auto const nclus = DSViter->size();
+      assert(nclus > 0);
+      clusInModule[gind] = nclus;
+      numberOfClusters += nclus;
+    }
+    hitsModuleStart[0] = 0;
+    assert(clusInModule.size() > gpuClustering::MaxNumModules);
+    for (int i = 1, n = clusInModule.size(); i < n; ++i)
+      hitsModuleStart[i] = hitsModuleStart[i - 1] + clusInModule[i - 1];
+    assert(numberOfClusters == int(hitsModuleStart[gpuClustering::MaxNumModules]));
+
+    // yes a unique ptr of a unique ptr so edm is happy and the pointer stay still...
+    iEvent.emplace(tHost_, std::move(hmsp));  // hmsp is gone, hitsModuleStart still alive and kicking...
```

Comment:
@mmusich @tsusa @VinInn
I would like to understand if these changes can be avoided ?

  - the legacy workflow (*e.g.* `11634.5`) does not use the `HostProduct<unsigned int[]>` product:
    ```python
    process.siPixelRecHitsPreSplitting = SwitchProducerCUDA(
        cpu = cms.EDProducer("SiPixelRecHitConverter",
            CPE = cms.string('PixelCPEGeneric'),
            VerboseLevel = cms.untracked.int32(0),
            src = cms.InputTag("siPixelClustersPreSplitting")
        )
    )
    ```
  - the SoA workflow (*e.g.* `11634.501`) does not run this module, using `SiPixelRecHitSoAFromLegacy` instead:
    ```python
    process.siPixelRecHitsPreSplitting = cms.EDProducer("SiPixelRecHitSoAFromLegacy",
        CPE = cms.string('PixelCPEFast'),
        beamSpot = cms.InputTag("offlineBeamSpot"),
        convertToLegacy = cms.bool(True),
        mightGet = cms.optional.untracked.vstring,
        src = cms.InputTag("siPixelClustersPreSplitting")
    )
    ```
  - the GPU workflow (*e.g.* `11634.501`) is the one that requires the extra product from `SiPixelRecHitConverter` to match the output of `SiPixelRecHitFromSOA`:
    ```python
    process.siPixelRecHitsPreSplitting = SwitchProducerCUDA(
        cpu = cms.EDProducer("SiPixelRecHitConverter",
            CPE = cms.string('PixelCPEGeneric'),
            VerboseLevel = cms.untracked.int32(0),
            src = cms.InputTag("siPixelClustersPreSplitting")
        ),
        cuda = cms.EDProducer("SiPixelRecHitFromSOA",
            mightGet = cms.optional.untracked.vstring,
            pixelRecHitSrc = cms.InputTag("siPixelRecHitsCUDAPreSplitting"),
            src = cms.InputTag("siPixelClustersPreSplitting")
        )
    )       
    ```

[On Feb 4](https://github.com/cms-patatrack/cmssw/pull/438/files#r374692709) @VinInn commented
> This producer shall produce the very same products of its "SoAonCPU&2Legacy" buddy: SiPixelRecHitSoAFromLegacy. Most probably it could produce an empty product.
> Or SiPixelRecHitSoAFromLegacy shall be split in two (or better everything merged in one, (actually fully refactorized)).
> See my presentation to the DPG code-transfer workshop.

I guess the presentation is [this one](https://indico.cern.ch/event/863657/contributions/3666696/attachments/1959745/3256622/PixelFullHeterogeneousWF.pdf).

But that doesn't make it clear to me why is the `HostProduct<unsigned int[]>` needed in the first place ?